include/hip/hcc_detail/program_state.hpp

```diff
@@ -558,92 +561,185 @@ const std::unordered_map<
+        // Check if this kernel was already processed
+        if(!kernargs[kernel_name].empty()) {
+            continue;
+        }
 
-            if (dx == std::string::npos) break;
+        comgr_metadata_node kernArgList;
+        status = amd_comgr_metadata_lookup(kernelMap.node, "Args", &kernArgList.node);
+        if (status == AMD_COMGR_STATUS_SUCCESS ) {
+            kernArgList.set_active(true);
+            status = amd_comgr_get_metadata_kind(kernArgList.node, &mkindLookup);
+            if (mkindLookup != AMD_COMGR_METADATA_KIND_LIST) {
+                 hip_throw(std::runtime_error{"Lookup of Args didn't return a list\n"});
+            }
 
-            static constexpr decltype(tmp.size()) name_sz{5};
-            dx = tmp.find_first_not_of(" '", dx + name_sz);
+            if (status == AMD_COMGR_STATUS_SUCCESS ) {
+                status = amd_comgr_get_metadata_list_size(kernArgList.node, &num_kern_args);
+                checkError(status, "amd_comgr_get_metadata_list_size");
+                for (int k_ar = 0; k_ar < num_kern_args; k_ar++) {
+                    comgr_metadata_node kernArgMap;
+                    status = amd_comgr_index_list_metadata(kernArgList.node, k_ar, &kernArgMap.node);
+                    checkError(status, "amd_comgr_index_list_metadata");
+                    kernArgMap.set_active(true);
+                    size_t k_arg_size, k_arg_align;
 
-            auto fn = tmp.substr(dx, tmp.find_first_of("'\n", dx) - dx);
-            dx += fn.size();
+                    k_arg_size = std::stoul(lookup_keyword_value(kernArgMap.node, "Size"));
+                    k_arg_align = std::stoul(lookup_keyword_value(kernArgMap.node, "Align"));
 
-            auto dx1 = tmp.find("CodeProps", dx);
-            dx = tmp.find("Args:", dx);
+                    // Save it into our kernargs
+                    kernargs[kernel_name].emplace_back(k_arg_size, k_arg_align);
 
-            if (dx1 < dx) {
-                dx = dx1;
-                continue;
+                } // end kernArgMap
             }
-            if (dx == std::string::npos) break;
+        } // end kernArgList
+    } // end kernelMap
+} // end kernelList
 
-            static constexpr decltype(tmp.size()) args_sz{5};
-            dx = parse_args(tmp, dx + args_sz, dx1, kernargs[fn]);
-        } while (true);
+inline
+void read_kernarg_metadata(
+    std::string blob,
+    std::unordered_map<
+        std::string,
+        std::vector<std::pair<std::size_t, std::size_t>>>& kernargs) {
+
+    const char *blob_buf = blob.data();
+    long blob_size = blob.size();
+
+    amd_comgr_status_t status;
+    comgr_data blob_data;
+    status = amd_comgr_create_data(AMD_COMGR_DATA_KIND_RELOCATABLE, &blob_data.data);
+    checkError(status, "amd_comgr_create_data");
+
+    status = amd_comgr_set_data(blob_data.data, blob_size, blob_buf);
+    if(status != AMD_COMGR_STATUS_SUCCESS)
+        return;
+    blob_data.set_active(true);
```

Comment:
I think this `set_active` should be moved before the call to `set_data`; a data object will leak if it can be created but cannot be set. I don't see any other issues with the use of comgr in the review.

Reply:
Thanks for catching that!